### PR TITLE
Android bugfixes and general improvements, as well as updates for Android 8 (Oreo).

### PR DESCRIPTION
**Pull request by knxrb**
**Description:**
Pull request for improvements and bugfixes I've come across while using this plugin in our app. Also adds support for Android 8 (Oreo), with the channels and badges, as well as specifying the colour for the notification in Android.

**Fixes:**
- #137
- #134 
- #127 
- Also fixes a bug in Android where tapping a local notification does nothing if Android has currently, or previously, grouped them together in the notifications tray area.

**My Changes (Android & Sample projects):**
- Added support for setting the notification channels in Android Oreo and above, with a default fallback channel, and support for enabling vibration, and showing the badge dot above the app icon.
- Added support for setting the colour for the Android notifications.
- Added a new option to the Android local notifications to make it attempt to bring focus back to the app when tapped. (Note: This gives best results when the MainLauncher Activity has a LaunchMode of SingleTop.)
- Added click Intent to Android scheduled local notifications to allow them to re-focus/open the app on tap as well.
- ~~Added two helpful debug options for Android; one for showing the internal notification id at the beginning of the notification title, and another to display a toast with the internal notification id and result (when a notification is tapped).~~
- Made the Android NotificationBuilder code more stable.
- Changed the notification result to return the Id as well as the Action.
- Updated the Target Framework Version to 8.0.
- Updated the Xamarin Forms version from 2.3.0.46-pre3 to 2.3.4.270.
- Updated the Sample app so it runs with newer versions of Visual Studio for Mac & VS 2017, and shows an example of setting the Android notification colour, etc.

_No changes have been made to iOS or UWP, other than the UWP csproj formats (so it loads in VS 2017)._

Hopefully this is useful, and if you have any suggestions, queries, or changes, please let me know.

### CODE REVIEW

**Comment by adamped** [comment]
thank you @knxrb, I was waiting for someone to come along and help :)

I have noticed all these bugs but been so busy lately with my job and other OSS/community stuff, it's not got off my todo list.

In the future, if you could break these up, into smaller PR's, one per feature, that would be easier for me to check, but I will go through and check this one as is.

File: Toasts.Forms.Plugin.Droid/AlarmHandler.cs [hunk 1]
```diff
@@ -3,33 +3,58 @@
...
-                     .SetAutoCancel(true);
+                .SetContentTitle(options.AndroidOptions.DebugShowIdInTitle ? "[" + id + "] " + options.Title : options.Title)
+                .SetContentText(options.Description)
+                .SetSmallIcon(options.AndroidOptions.SmallDrawableIcon.Value) // Must have small icon to display
+                .SetPriority((int)NotificationPriority.High) // Must be set to High to get Heads-up notification
+                .SetDefaults(NotificationDefaults.All) // Must also include vibrate to get Heads-up notification
+                .SetAutoCancel(true)
+                .SetColor(Color.ParseColor(options.AndroidOptions.HexColour));
+
+            if (options.AndroidOptions.ForceOpenAppOnNotificationTap)
```
Comment (adamped) [inline Toasts.Forms.Plugin.Droid/AlarmHandler.cs]: Note to self to check this. By default it should open the notification. Need to make sure this is default, and this option is here to turn off the feature if they desire. Also I try to keep options the same over every platform. Will need to see if this is possible on iOS and UWP.

File: Toasts.Forms.Plugin.Droid/NotificationBuilder.cs [hunk 2]
```diff
@@ -1,95 +1,145 @@
...
             _receiver = new NotificationReceiver();
 
-            activity.RegisterReceiver(_receiver, filter);
+            Application.Context.RegisterReceiver(_receiver, filter);
 
             _androidOptions = androidOptions;
         }
 
-        public IList<INotification> GetDeliveredNotifications(Activity activity)
+        public static string GetOrCreateChannel(IAndroidChannelOptions channelOptions)
```
Comment (adamped) [inline Toasts.Forms.Plugin.Droid/NotificationBuilder.cs]: Why static?

Comment (knxrb) [reply Toasts.Forms.Plugin.Droid/NotificationBuilder.cs]: That is static as the AlarmHandler uses it as well to  create the channel for scheduled notifications.

File: Toasts.Forms.Plugin.Abstractions/Options/AndroidOptions.cs [hunk 3]
```diff
@@ -10,5 +12,12 @@ public class AndroidOptions : IAndroidOptions
         public int? SmallDrawableIcon { get; set; } = null;
         public string DismissText { get; set; } = "Dismiss";
         public string ViewText { get; set; } = "View";
+        public string HexColour { get; set; } = "#FFFFFFFF";
+        public bool ForceOpenAppOnNotificationTap { get; set; } = false;
+        public IAndroidChannelOptions ChannelOptions { get; set; } = new AndroidChannelOptions();
+
+        // Debug Help
+        public bool DebugShowCallbackToast { get; set; } = false;
+        public bool DebugShowIdInTitle { get; set; } = false;
```
[the remark this answers is not in the document]
Comment (knxrb) [reply Toasts.Forms.Plugin.Abstractions/Options/AndroidOptions.cs]: No problem, I'll delete these for now.

File: Toasts.Forms.Plugin.Abstractions/Interfaces/IAndroidOptions.cs [hunk 4]
```diff
@@ -1,12 +1,52 @@
...
         string ViewText { get; set; }
         string DismissText { get; set; }
+
+        /// <summary>
+        /// Gets or sets the hex colour for the notification to use.
+        /// Colour usage changes based on the Android version.
+        /// Example: #FF00CC
+        /// </summary>
+        /// <value>The hex colour</value>
+        string HexColour { get; set; }
```
[the remark this answers is not in the document]
Comment (knxrb) [reply Toasts.Forms.Plugin.Abstractions/Interfaces/IAndroidOptions.cs]: Will do, makes sense.

**Comment by adamped** [comment]
@knxrb - was merged in, but I did have to make a number of changes.

1. I added the AllowTapInNotificationCenter property. This has to be a new property, otherwise it would have broken people's existing implementations.
2. Added the same functionality to iOS and UWP
3. Added badge option to iOSOptions

Thanks for the PR.

**Pull request merged**